Add a workflow selection strategy for classrooms

eatyourgreens opened the pull request:
Add a new workflow strategy for classrooms, which always loads the workflow from a workflow query parameter.

ToDo:

- [x] Conditional logic to choose the selection strategy
- [x] Tests

Staging branch URL: https://pr-5259.pfe-preview.zooniverse.org

Fixes #5256.

Describe your changes.

# Required Manual Testing

- [ ] Does the non-logged in home page render correctly?
- [ ] Does the logged in home page render correctly?
- [ ] Does the projects page render correctly?
- [ ] Can you load project home pages?
- [ ] Can you load the classification page?
- [ ] Can you submit a classification?
- [ ] Does talk load correctly?
- [ ] Can you post a talk comment?

# Review Checklist

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?
- [ ] If the component is in coffeescript, is it converted to ES6? Is it free of eslint errors? Is the conversion its own commit?
- [ ] Are the tests passing locally and on Travis?

# Optional

- [ ] Have you replaced any `ChangeListener` or `PromiseRenderer` components with code that updates component state?
- [ ] If changes are made to the classifier, does the dev classifier still work?
- [ ] Have you [resized and compressed](https://developers.google.com/web/fundamentals/performance/optimizing-content-efficiency/image-optimization) any image you've added?
- [ ] Have you added in [flow type annotations](https://flowtype.org/docs/type-annotations.html)?
- [ ] Have you followed the [Springer guidelines for commit messages](https://github.com/springernature/frontend-playbook/blob/master/git/git.md#commit-messages)?

eatyourgreens commented:
@shaunanoordin all this should need now is a conditional to decide when to use the new strategy.

shaunanoordin commented:
## Discussion Notes

Some quick documentation of the real-life discussion Jim & me had.

(Context for people jumping in: read #5256, basically WildCam Lab needs the ability to choose workflows (i.e. classroom assignments) that are inactive, which requires modifying the PFE workflow selection logic, which would be best served by creating a Worklow Selection Strategy _Selector._ )

- Point: the conditional logic to check if we should use the "Classroom's workflow selection strategy" could just check if `?classroom=yes` is in the query.
- Counterpoint: wait, no, bad idea, this means any Project will let user bypass standard workflow selection by passing in a query.

- Point: instead of specifying `?workflow=1234&classroom=yup`, we could just use `?classroom=1234`
- Counterpoint: see above - if we're not checking the query to see if we're in a classroom or not, then this is a bit pointless.

- Point: we should be wary about letting users access workflows they're not supposed to, in PFE.
- Counterpoint: do we though? @marten pointed out that if we wanted to be really strict about workflow access, security needs to be placed on the backend, not the front end, where anyone with a modicum of JS skills can hack things.

- Point: it'll make more sense for the conditional logic to check if a Project is _configured_ to use the "Classroom Workflow Selection Strategy"
- Counterpoint: better figure out whether you want to put that config in `project.config` or `project.experimental_tools`
- Answer: I say `experimental_tools` for now.

- Point: this new Workflow Selection Strategy philosophy should be implemented to reduce overloading an already fragile Workflow Selection (see `workflow-selection.jsx`) logic.
- Counterpoint: in that case, better check that [front-end-monorepo](https://github.com/zooniverse/front-end-monorepo) is already thinking about this.

- Point: 👉 
- Counterpoint: 👈

eatyourgreens commented:
@shaunanoordin will the classroom projects also allow volunteers to choose a workflow from the home page? Things get a bit more complicated if the new component is also used by regular volunteers.

In an ideal world, the `ClassroomWorkflowSelection` component would only be used by students doing classroom work.

shaunanoordin commented:
> will the classroom projects also allow volunteers to choose a workflow from the home page? Things get a bit more complicated if the new component is also used by regular volunteers

Fortunately for this case, WildCam Lab's assignments create workflows that are _inactive,_ so 1. they won't appear on the home page even if allowed the standard "Users Can Choose Workflow" option (tested yesterday), and 2. Classifications on the workflow won't accelerate retirement.

eatyourgreens commented:
Thanks @shaunanoordin. Will there ever be a case where there are two or more active workflows on the home page, for regular volunteers, plus a bunch of inactive workflows for students?

In that case, we'd have to write logic to use regular selection for the home page choices but classroom selection for the students.

shaunanoordin commented:
Note: at the moment, I'm already looking at some code logic for the conditional check.

Solution 1A: check if the Project has the `wildcam classroom` Experimental Tool.

`const isClassroom = (props.project && props.project.experimental_tools && props.project.experimental_tools.indexOf('wildcam classroom') > -1);` works for [WildCam Darien](https://www.zooniverse.org/projects/wildcam/wildcam-darien/) (which has `'wildcam classroom'` in its experimental_tools config)... but only if a `?workflow` is specified. Solution 1A breaks for standard volunteers attempting to work on the default workflow, because it passes standard voluneers to Classroom Workflow Selection strategy even if they're not students.

So...

Solution 1B: check if the Project has the `wildcam classroom` Experimental Tool AND there's `classroom=1` in the query

Working on this.

eatyourgreens commented:
@shaunanoordin I guess the inverse of that would also be that classrooms can't be added to projects that have more than one active workflow.

eatyourgreens commented:
NB. [ES6 super() syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/super) would allow us to fall back by calling `super.getSelectedWorkflow(props)` but I'm not sure if that's the desired behaviour for classroom students. It would definitely work for regular volunteers.

eatyourgreens commented:
@shaunanoordin I'm leaning towards thinking that the logic should be: if the project has classrooms AND the URL looks like a classroom URL (has a workflow param, maybe has a classroom param too?)

We could update the PFE routes to map `/projects/{project_slug}/classroom` to the classifier, then use the new strategy on that route. It depends whether changing the classroom links is too big of a change.

shaunanoordin commented:
@eatyourgreens Changing the classroom links, at Zooniverse Classrooms (aka Education API Front End) ain't a problem whatsoever: https://github.com/zooniverse/edu-api-front-end/blob/master/src/programs/darien/wildcam-darien.classroom-config.js

eatyourgreens commented:
By the way, I set up the fallback here to return null, rather than retrying and loading the default workflow, because I assumed that a student would not want to start doing their classroom assignment only to realise they were on the active workflow.

shaunanoordin commented:
I'm about to push some changes to this PR that will address the _conditional check._

```
const isClassroom = true; 
---- change to -->
//Check for WildCam Lab classrooms (see https://github.com/zooniverse/edu-api-front-end)
const isProjectForClassrooms = (props.project && props.project.experimental_tools && props.project.experimental_tools.indexOf('wildcam classroom') > -1);
const isUrlForClassrooms = props.location.query && props.location.query.classroom;
const isClassroom = isProjectForClassrooms && isUrlForClassrooms;

```

- a Classroom URL is only valid if 1. the Project is configured to accept classrooms and 2. the URL has the `?classrooms=yup` in its query
- if the Classroom URL is invalid, the Classifier uses the standard workflow selection logic.
- if the Classroom URL is invalid but the worklow ID is invalid (e.g. deleted workflow), then the Classifier displays blank content - THIS IS INTENDED. For Classrooms, it's better to error out than to direct students to an incorrect workflow (i.e. the default workflow)

External changes required:
- WildCam Darien Lab on Education API Front End needs to changes its Assignment URL template (Shaun to make changes)
- WildCam Darien needs to have `'wildcam classroom'` added to its Experimental Tools (done ✔️)

eatyourgreens commented:
@shaunanoordin will the changes here affect the astronomy classrooms too?

eatyourgreens commented:
Looks good to me but I will give it a closer look tomorrow.

eatyourgreens commented:
@shaunanoordin I'm going to make one small change, which is to add a check that there is a workflow present in the URL.

eatyourgreens commented:
@shaunanoordin in the last example, `workflow=11111111` the page is blank because it enters an infinite loop of constantly requesting the invalid workflow from Panoptes. I thought returning null would override the retry but obviously not. I'll look into this.

eatyourgreens commented:
@shaunanoordin this could do with another look. I've rewritten the code to avoid the infinite loop, which happened because `getWorkflow` will call `getSelectedWorkflow` if it can't find the workflow ID in the linked workflows for the project. `getSelectedWorkflow` will then call `getWorkflow` with the invalid workflow from the URL, so `getWorkflow` will call `getSelectedWorkflow` etc. etc. etc. I've updated the new component to call the `loadWorkflow` action just once when the component mounts.

I've also updated the tests and removed unused code that was leftover from the original tests.

eatyourgreens commented:
I think this demonstrates that small components, and simple functions that do one thing well, are a lot easier to work with than a complex component with functions that try to handle every possible case for workflow selection.

shaunanoordin commented:
👍 Cool, thanks Jim - give me a few minutes to run through my tests

eatyourgreens commented:
I changed the rule for classrooms to look for `classroom` and `workflow` in the URL, but that can be changed back.

srallen commented:
Sorry can someone please explain to me why the workflows couldn't just be active to begin with? None of the discussion above clearly states why they were inactive. My suggestion would have been to have the workflows be set to `active: true` when created by Edu API (by POSTING to Panoptes) and set them to be hidden from the stats page (a workflow config). It is my guess that they are inactive because they didn't want to be seen on the stats page of the project. This would have been a more simple fix that didn't require a new workflow selection strategy.

marten commented:
Wouldn't that mean they'd get randomly assigned to normal volunteers?

eatyourgreens commented:
@srallen my understanding was classrooms don't retire subjects, so the classifications aren't counted.

marten commented:
Oh, also if you make them active, they'll all appear in the `workflow.links.active_workflows` field when requested from the API. And it's a loooot of workflows.

srallen commented:
The Wildcam project has a default workflow? Retirement can be set to "never_retire" independently of activeness, which is what the I2A classrooms do. The serialization is a good point, though.

shaunanoordin commented:
@srallen Sorry, the original rationale was buried here in the issue #5256 :

> The workflows created by WildCam Assignments are inactive, to prevent Assignments-based Classifications from accidentally retiring Subjects too early. 

The original WildCam system was designed to keep the 'official' workflow separate from the school-created ones, since setting teacher-created workflows to 'active' may have a number of unwanted side effects to the official research.

Also - Darien's official workflow: 3033, Gorongosa's official workflow: 338.

eatyourgreens commented:
Darién has a default workflow. One other thing we wanted to do was override default error handling, so that students don't classify on the default workflow by mistake.

srallen commented:
The assignment workflows retirement can be set to `never_retire`. It's possibly Wildcam was setup before that retirement strategy was supported by Panoptes, but it can be used now. 

There's a banner that shows up for the query param for `group` used by I2A which validates that the user is a member of that user group. All classrooms are just Panoptes user groups. I had suggested that Wildcam use this for classroom membership validation as well and so that that UI component shows up. 

We now have two different ways that classroom workflows are rendered.

marten commented:
(Retirement is set to that for all workflows EduAPI makes, and has been that way from basically the start, so has always applied to WG:OG as well as Darien)

srallen commented:
Ok, so if retirement is set to `never_retire` then they could be set to be active? That isn't a reason to keep them inactive as far as I can tell. The serialization on the project is the only potential blocker I can see.

camallen commented:
We currently serialize all the workflows anyway via links, changing them to `active` would result in another probably expensive serializer query though. 
<img width="810" alt="screenshot 2019-02-08 at 15 41 38" src="https://user-images.githubusercontent.com/295329/52488589-10f55080-2bb8-11e9-81bf-d963e4922331.png">

Personally i prefer the strategy pattern to simplify the logic of the workflow selection. The default selection code is burdened with logical flows that are hard to understand and debug.

srallen commented:
Yeah, I understand the serialization concern. The strategy approach is what we were planning on for the rewrite. We need to minimize changes here and we had a potential solution that did not involve modifying PFE code.

We now have two different ways that the workflows for classroom associated projects are being rendered and it's not clear which does what. Does Wildcam projects validate that the logged in user is a member of the group aka classroom? Does it POST classifications with the user group id in the metadata like I2A too (and is this necessary)? 

For the Edu API, we do have a distinction between "Wildcam like" and "Intro 2 Astro like" education programs, if there are truly differing requirements for the rendering of the workflow, then it should be documented. If not, then there should be consistency for projects that are associated with the Edu API.
